From 88febbaba9036c46119ad721f9948553e48b65d0 Mon Sep 17 00:00:00 2001
From: Liu Zixian <liuzixian4@huawei.com>
Date: Fri, 17 Sep 2021 15:16:02 +0000
Subject: [PATCH 4/4] support vm_mmap_hugepages in ksys_mmap_pgoff

Signed-off-by: Liu Zixian <liuzixian4@huawei.com>
---
 mm/mmap.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/mm/mmap.c b/mm/mmap.c
index 386c3ab5a..7c73f5779 100644
--- a/mm/mmap.c
+++ b/mm/mmap.c
@@ -1632,6 +1632,16 @@ unsigned long ksys_mmap_pgoff(unsigned long addr, unsigned long len,
 		struct user_struct *user = NULL;
 		struct hstate *hs;
 
+#ifdef CONFIG_EULEROS_EXEC_HUGEPAGES
+		if (fd > 0) {
+			file = fget(fd);
+			if (file) {
+				retval = vm_mmap_hugepage(file, addr, len,
+						prot, flags, pgoff << PAGE_SHIFT);
+				goto out_fput;
+			}
+		}
+#endif
 		hs = hstate_sizelog((flags >> MAP_HUGE_SHIFT) & MAP_HUGE_MASK);
 		if (!hs)
 			return -EINVAL;
-- 
2.27.0

